File: packages/jexlang-java/src/main/java/com/jexlang/java/types/JexDouble.java
package com.jexlang.java.types;

public class JexDouble implements JexValue {
    private final Double value;
    public JexDouble(Double value) { this.value = value; }

    @Override
    public String getType() {
        return "double";
    }

    @Override
    public String toString() {
        return String.format("%.2f", value);
    }

    @Override
    public Object toObject() {
        return value; // Represents the double value in Java
    }

    public boolean isInteger() { return false; }
    public boolean isDouble() { return true; }
    public boolean isNumber() { return false; }
    public boolean isBoolean() { return false; }
    public boolean isString() { return false; }
    public boolean isNull() { return false; }
    public boolean isArray() { return false; }
    public boolean isObject() { return false; }
    public Integer asInteger(String ctx) { throw JexValue.typeError("integer", ctx, this); }
    public Double asDouble(String ctx) { return value; }
    public Number asNumber(String ctx) { throw JexValue.typeError("number", ctx, this); }
    public boolean asBoolean(String ctx) { throw JexValue.typeError("boolean", ctx, this); }
    public String asString(String ctx) { throw JexValue.typeError("string", ctx, this); }
    public java.util.List<JexValue> asArray(String ctx) { throw JexValue.typeError("array", ctx, this); }
    public java.util.Map<String, JexValue> asObject(String ctx) { throw JexValue.typeError("object", ctx, this); }
}

File: packages/jexlang-swift/jexlang-swift/eval/types/JexDouble.swift
//
//  JexInteger.swift
//  jexlang-swift
//
//  Created by Vinay Kumar on 03/10/25.
//

import Foundation

public class JexDouble: JexValue {
    
    private let value: Double
    
    public init(value: Double) {
        self.value = value
    }
    
    public func isInteger() -> Bool {
        return false
    }
    
    public func isDouble() -> Bool {
        return true
    }
    
    public func isNumber() -> Bool {
        return false
    }
    
    public func isBoolean() -> Bool {
        return false
    }
    
    public func isString() -> Bool {
        return false
    }
    
    public func isNil() -> Bool {
        return false
    }
    
    public func isArray() -> Bool {
        return false
    }
    
    public func isObject() -> Bool {
        return false
    }
    
    public func getType() -> String {
        return "double"
    }
    
    public var description: String {
        return "\(value)"
    }
    
    public func toObject() -> AnyObject? {
        return value as AnyObject;
    }
    
    public func asDouble(context: String) -> Double {
        return value
    }
    
    public func asInteger(context: String) throws -> Int {
        throw JexValueFactory.typeError(want: "integer", ctx: context, actualValue: self)
    }
    
    public func asNumber(context: String) throws -> NSNumber {
        throw JexValueFactory.typeError(want: "number", ctx: context, actualValue: self)
    }
    
    public func asBoolean(context: String) throws -> Bool {
        throw JexValueFactory.typeError(want: "boolean", ctx: context, actualValue: self)
    }
    
    public func asString(context: String) throws -> String {
        throw JexValueFactory.typeError(want: "string", ctx: context, actualValue: self)
    }
    
    public func asArray(context: String) throws -> [JexValue] {
        throw JexValueFactory.typeError(want: "array", ctx: context, actualValue: self)
    }
    
    public func asObject(context: String) throws -> [String : JexValue] {
        throw JexValueFactory.typeError(want: "object", ctx: context, actualValue: self)
    }
    
    public func equals(_ other: JexValue) -> Bool {
        if other.isDouble() {
            return (try? other.asDouble(context: "equals")) == value
        }
        if other.isNumber() {
            return (try? other.asNumber(context: "equals")).map { $0.doubleValue == value } ?? false
        }
        if other.isInteger() {
            return (try? other.asInteger(context: "equals")).map { Double($0) == value } ?? false
        }
        return false
    }
}

